### ISSUE TITLE
Add action for removing attachments 

### ISSUE THREAD

**Issue by tmalsburg** [djcb/mu#866]
I often receive emails with large attachments. My email provider ([Posteo](http://www.posteo.de)) charges per GB and usually I only want to keep the email but not the attachments anyway.  Therefore, it would be useful if Mu4e had a command for removing attachments. @mankoff posted some code for doing this in #790 and it's working well so far.

Some details that should be considered:

1. @mankoff's solution requires the altermine tool which may not be available on the average user's system.
2. It would be great if the attachment was not just deleted but rather be replaced by the name of the formerly attached file (altermine can apparently do this, see [here](https://github.com/djcb/mu/issues/790#issuecomment-224748650) for details).
3. If an email has many attachments, it can be tedious to delete them all individually.  It would be food to have the ability to delete multiple/all attachments (similar to `C-u e` for extracting all attachments).

**Comment by djcb** [comment]
This would require a bit of new machinery in `mu extract`; in addition, `mu` is not really a fan of modifying messages in the first place. 

So, this is probably best left to altermime and similar tools; if features are missing, that might be a good place to ask for them.

**Comment by tmalsburg** [comment]
I don't see why this needs to be handled in mu.  Ken's solution bypasses mu and edits the messages in the Maildir directly.  Also altermime already has all features that we need, it would just be nice to have a front-end for that in Mu4e because it would be inconvenient to remove every single attachment by calling altermime manually in a shell.  Am I misunderstanding something?

**Comment by mankoff** [comment]
Just to give credit where due, the altermime code comes from @djcb, and is not "Ken's solution", as posted both here and #790

**Comment by tmalsburg** [comment]
Ah, sorry, I missed that bit.  Thanks for clarifying, Ken.

**Comment by djcb** [comment]
From the description above (point 1) I got that this is about a mu-reimplementation of the altermime functionality... and I explained why I don't want to go that path. But if you want to extend @mankoff 's altermime-based action, by all means, go ahead!

**Comment by tmalsburg** [comment]
You closed the issue, from that I inferred that you don't want this feater at all.  Perhaps that was the misunderstanding.

Regarding the implementation: what about a reimplementation in Mu4e (pure Elisp) leaving mu untouched. I don't see why this needs to be in the indexer.

Would you also accept to altermime solution even though it may not work out of the box for users who don't have altermime installed?

**Comment by djcb** [comment]
I closed it because it wouldn't happen given the considerations mentioned; of course, implementing it as an action calling e.g. altermime is fine. Implementing that in elisp _without_ altermime would be a lot of work...

**Comment by djcb** [comment]
(oh, I'd be happy to re-open the issue if someone wants to work on this)

**Comment by gambhiro** [comment]
Thanks for the useful info here and in #790!

Starting from this, I started a package to "detach" attachments, i.e. to save and remove them.

https://github.com/gambhiro/mu4e-detach

Saving works good and `altermime` removes them from the message file.

Do you know how could I sync the changes back with offlineimap? Since the message UID doesn't change, it doesn't update the message contents.

I believe what the Thunderbird "detach" action does is that it creates a copy of the message (all the same, but without attachments) and removes the original, then this is synced to the server.

Can I do that with the mu4e API, to create a new copy of the altered message?

**Comment by tmalsburg** [comment]
@gambhiro, thanks for working on this and making your code available.  Have you considered making a PR for mu4e?  I think a log of users might benefit from this feature and some may not discover it when it lives in a separate package.

**Comment by gambhiro** [comment]
@tmalsburg at the moment it is missing an important function -- messages don't get updated on the server. The message UID doesn't change, and so offlineimap sees no change.

What I think has to happen is that the altered message has to be copied as a new message, and the original removed, then offlineimap would sync that.

I don't know how to go about that actually.

It is a fairly destructive procedure as well. So I thought it is best if the default behaviour is to not remove the attachments, and the user has to manually add the setting to enable that.

Once it actually works, and @djcb thinks it is suitable, I would be glad to make it a PR.

**Comment by tmalsburg** [comment]
I agree that removing attachments should definitely not be a default behavior.  I think this is rather something that users would trigger manually for individual messages that consume too much disk space.  Propagation to the server is of course important because that's were the cost is incurred (at least when a paid email service is used).

Does a changed UID imply that org-links to the message would break?  That would be unfortunate.

**Comment by gambhiro** [comment]
Good point about the org links.

I wouldn't want them them to break either, but it should be easy to collect a list of old and new UIDs, then replace them in all links.

There is also a lame solution in my code which should be improved.

When the message is retreived by `mu4e~proc-send-command`, the function registered as `mu4e-view-func` receives it. Normally this is `mu4e~headers-view-handler`. So to get the full message data in one of our functions, this has to be redefined before we start everything. And restored after we are done, it has to be restored to `mu4e~headers-view-handler`.

It figures out when to restore the view handler by incrementing and decrementing a simple counter, which works but it's a bit painful to see it work like that.

It would be better to have a hook in mu4e, before all marks are processed and after all marks are done.

**Comment by tmalsburg** [comment]
> I wouldn't want them them to break either, but it should be easy to collect a list of old and new UIDs, then replace them in all links.

This assumes that we have a complete list of all org documents containing such links.  At least in my set-up that's not the case.  A solution preserving the links would be preferable. Perhaps the authors of offlineimap and mbsync (the other syncer) can provide input on how to update remove copies without changing the UID?

Re implementation: I'm not terribly familiar with mu4e internals, but perhaps @djcb can give input.

**Comment by mankoff** [comment]
This is an old thread, but just pointing people that read it to https://github.com/djcb/mu/issues/2247 where an improved `remove-attachment` function leaves the name of the removed file, and there is a `remove-all-attachments` function.
